### PR TITLE
Customizing for small slits and bright standards

### PR DESCRIPTION
**Pull request by profxj**
A small set of updates to get 2x2 binning with MagE to reduce 
and flux.  Includes a number of "bug fixes" to `fluxspec`. 
Here are the main changes:

- Add find_trim_edge and std_prof_nsigma parameters
- A bit of tuning for MagE
- Fixes for Echelle in fluxspec

### CODE REVIEW

File: pypeit/fluxspec.py [hunk 1]
```diff
@@ -444,22 +447,27 @@ def generate_sensfunc(self):
             return None
 
         ext_final = fits.getheader(self.par['std_file'], -1)
-        norder = ext_final['ECHORDER'] + 1
```
Comment (feigewang) [inline pypeit/fluxspec.py]: this might be tricky if you have two objects in the standard 1d spectra file

File: pypeit/fluxspec.py [hunk 2]
```diff
@@ -444,22 +447,27 @@ def generate_sensfunc(self):
...
+        #norder = ext_final['ECHORDER'] + 1
+        norder = len(self.std)
```
Comment (feigewang) [inline pypeit/fluxspec.py]: this might be tricky if you have two objects in the standard 1d spectra file

Comment (profxj) [reply pypeit/fluxspec.py]: My change ignores the spectra file.
It keys on `self.std` which has been constructed
to be only a single object.   Do you agree?

**Comment by profxj** [comment]
Passing all unit tests:

(base) profx> py.test --disable-warnings 
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-4.3.1, py-1.8.0, pluggy-0.9.0
rootdir: /data/Projects/Python/PypeIt, inifile:
plugins: remotedata-0.3.1, openfiles-0.3.2, doctestplus-0.3.0, cov-2.7.1, arraydiff-0.3
collected 169 items                                                            

pypeit/tests/test_arc.py .                                               [  0%]
pypeit/tests/test_biasframe.py ....                                      [  2%]
pypeit/tests/test_bpmimage.py ...                                        [  4%]
pypeit/tests/test_calibrationimage.py .                                  [  5%]
pypeit/tests/test_calibrations.py ..........                             [ 11%]
pypeit/tests/test_coadd.py ...........                                   [ 17%]
pypeit/tests/test_cooked.py .                                            [ 18%]
pypeit/tests/test_flatfield.py .                                         [ 18%]
pypeit/tests/test_flexure.py .                                           [ 19%]
pypeit/tests/test_flux.py ....                                           [ 21%]
pypeit/tests/test_fluxspec.py ...                                        [ 23%]
pypeit/tests/test_load.py ..                                             [ 24%]
pypeit/tests/test_load_images.py ..........                              [ 30%]
pypeit/tests/test_masterframe.py .                                       [ 31%]
pypeit/tests/test_match.py ....                                          [ 33%]
pypeit/tests/test_metadata.py .....                                      [ 36%]
pypeit/tests/test_msgs.py ..                                             [ 37%]
pypeit/tests/test_opticalmodel.py ....                                   [ 40%]
pypeit/tests/test_parse.py ...                                           [ 42%]
pypeit/tests/test_processimage.py ..                                     [ 43%]
pypeit/tests/test_procimg.py .                                           [ 43%]
pypeit/tests/test_pydl.py .                                              [ 44%]
pypeit/tests/test_pypeit.py ..                                           [ 45%]
pypeit/tests/test_pypeitimage.py ..                                      [ 46%]
pypeit/tests/test_pypeitpar.py ......................                    [ 59%]
pypeit/tests/test_pypeitsetup.py .......                                 [ 63%]
pypeit/tests/test_qa.py .                                                [ 64%]
pypeit/tests/test_save.py ..                                             [ 65%]
pypeit/tests/test_scienceimage.py ....                                   [ 68%]
pypeit/tests/test_scripts.py ......                                      [ 71%]
pypeit/tests/test_setups.py .............                                [ 79%]
pypeit/tests/test_slitmask.py .                                          [ 79%]
pypeit/tests/test_specobj.py ..                                          [ 81%]
pypeit/tests/test_specobjs.py ....                                       [ 83%]
pypeit/tests/test_spectrographs.py ............                          [ 90%]
pypeit/tests/test_traceimage.py ..                                       [ 91%]
pypeit/tests/test_traceslits.py .....                                    [ 94%]
pypeit/tests/test_utils.py ..                                            [ 95%]
pypeit/tests/test_vcorr.py ..                                            [ 97%]
pypeit/tests/test_wavecalib.py .                                         [ 97%]
pypeit/tests/test_waveimage.py .                                         [ 98%]
pypeit/tests/test_wavemodel.py .                                         [ 98%]
pypeit/tests/test_wavetilts.py ..                                        [100%]

================== 169 passed, 162 warnings in 763.95 seconds ==================

File: pypeit/fluxspec.py (hunk 2 above)
Comment (feigewang) [reply pypeit/fluxspec.py]: okay. I forgot that we have already run the find_standard before this step.

File: pypeit/core/skysub.py [hunk 3]
```diff
@@ -618,7 +618,7 @@ def local_skysub_extract(sciimg, sciivar, tilts, waveimg, global_sky, rn2_img, t
     if std is True:
         chi2_sigrej = 100.0
         #sigrej_ceil = 1e10
-        sigrej = 25.0
+        sigrej = 50.0  # 25 wasn't enough for MagE 2x2 binning (probably undersampled)
```
Comment (jhennawi) [inline pypeit/core/skysub.py]: Maybe we should turn this off entirely?

File: pypeit/fluxspec.py [hunk 4]
```diff
@@ -283,7 +279,14 @@ def show_sensfunc(self):
         #mag_func = utils.func_val(self.sens_dict['c'], wave, self.sens_dict['func'])
         #sens = 10.0**(0.4*mag_func)
         # Plot
-        debugger.plot1d(self.sens_dict['wave'], self.sens_dict['sensfunc'], xlbl='Wavelength', ylbl='Sensitivity Function')
+        if self.spectrograph.pypeline == 'Echelle':
```
Comment (jhennawi) [inline pypeit/fluxspec.py]: Can you please get the old debugger out of here, it is deprecated.

File: pypeit/fluxspec.py [hunk 5]
```diff
@@ -362,7 +365,7 @@ def generate_sensfunc(self):
 
         self.sens_dict = {}
         try:
-            this_wave = self.std.optimal['WAVE_GRID']
+            this_wave = self.std.boxcar['WAVE_GRID']
         except KeyError:
```
Comment (jhennawi) [inline pypeit/fluxspec.py]: I'm not convinced this makes a difference, and would like you to show a plot of both optimal and boxcar showing the differences before changing this.

File: pypeit/fluxspec.py [hunk 6]
```diff
@@ -444,22 +447,27 @@ def generate_sensfunc(self):
...
-            std_specobjs, std_header = load.load_specobjs(self.par['std_file'], order=iord)
-            std_idx = flux.find_standard(std_specobjs)
-            std = std_specobjs[std_idx]
+            #std_specobjs, std_header = load.load_specobjs(self.par['std_file'], order=iord)
+            #embed(header='447 of fluxspec')
+            #std_idx = flux.find_standard(std_specobjs)
+            std = self.std[iord] #std_specobjs[std_idx]
+
+            # THIS IS A CRAZY KLUDGE....
             try:
```
Comment (jhennawi) [inline pypeit/fluxspec.py]: wave_grid is going to be depcreataed, this can be removed.

File: pypeit/fluxspec.py [hunk 7]
```diff
@@ -444,22 +447,27 @@ def generate_sensfunc(self):
...
                 wavemask = std.optimal['WAVE_GRID'] > 0.0 #*units.AA
             except KeyError:
-                wavemask = std.optimal['WAVE'] > 1000.0 * units.AA
-                this_wave = std.optimal['WAVE'][wavemask]
+                wavemask = std.boxcar['WAVE'] > 1000.0 * units.AA
```
Comment (jhennawi) [inline pypeit/fluxspec.py]: I would prefer to not have quantities anywhere in PypeIt.

File: pypeit/par/pypeitpar.py [hunk 8]
```diff
@@ -1743,6 +1750,10 @@ def __init__(self, bspline_spacing=None, boxcar_radius=None, trace_npoly=None,
         descr['maxnumber'] = 'Maximum number of objects to extract in a science frame.  Use ' \
                              'None for no limit.'
 
+        defaults['find_trim_edge'] = [5,5]
```
Comment (jhennawi) [inline pypeit/par/pypeitpar.py]: The description is incorrect. It does not trim the slit. It simply ignores objects found within this many pixels of the slit edge. Also, we should probably consider making this arc seconds.

File: pypeit/par/pypeitpar.py [hunk 9]
```diff
@@ -1743,6 +1750,10 @@ def __init__(self, bspline_spacing=None, boxcar_radius=None, trace_npoly=None,
...
+        dtypes['find_trim_edge'] = list
```
Comment (jhennawi) [inline pypeit/par/pypeitpar.py]: Shouldn't this be a tuple? It is not mutable.

File: pypeit/core/skysub.py (hunk 3 above)
Comment (profxj) [reply pypeit/core/skysub.py]: Sounds dangerous.  Am going to make that an Issue #717 but
leave it for now.

File: pypeit/fluxspec.py (hunk 4 above)
Comment (profxj) [reply pypeit/fluxspec.py]: The old debugger is not deprecated.
Only its wrapper to `pdb`.  Removing that now.

File: pypeit/fluxspec.py (hunk 5 above)
Comment (profxj) [reply pypeit/fluxspec.py]: ok, am moving back for now.

File: pypeit/fluxspec.py (hunk 6 above)
Comment (profxj) [reply pypeit/fluxspec.py]: ok, edited.

File: pypeit/fluxspec.py (hunk 7 above)
Comment (profxj) [reply pypeit/fluxspec.py]: am hesitant to remove this now.
you can do so as you refactor the fluxing.

File: pypeit/par/pypeitpar.py (hunk 8 above)
Comment (profxj) [reply pypeit/par/pypeitpar.py]: Look at this again, @jhennawi 

see this line in `extract.py`

```
    ximg, edgmask = pixels.ximg_and_edgemask(slit_left, slit_righ, thismask, trim_edg=trim_edg)
```

which I think has the effect of zeroing out the slit in these trimmed edges.

File: pypeit/par/pypeitpar.py (hunk 9 above)
Comment (profxj) [reply pypeit/par/pypeitpar.py]: I tried and failed.  Will try again.

File: pypeit/par/pypeitpar.py (hunk 8 above)
Comment (jhennawi) [reply pypeit/par/pypeitpar.py]: It does other things with trim_edge. Look in the code;).

File: pypeit/par/pypeitpar.py (hunk 9 above)
Comment (profxj) [reply pypeit/par/pypeitpar.py]: sticking with list for now.

**Comment by profxj** [comment]
merging.

**Pull request merged**